src/ledger/user.js

```diff
@@ -0,0 +1,105 @@
+// @flow
+
+/**
+ * This module has a core data type identifying SourceCred users.
+ *
+ * The scope for this data type is to model:
+ * - a unique identifier for each user
+ * - a unique (renameable) username they choose
+ * - the address of every node they correspond to in the graph
+ *
+ * Unlike most other state in SourceCred, the User state is
+ * nondeterministically generated by SourceCred itself, and then persisted
+ * long-term within the instance.
+ *
+ * This is in contrast to Graph data, which usually comes from an external
+ * source, and is not persisted long-term, but instead is re-generated when
+ * needed.
+ *
+ * In particular, this kernel of user data is stored within the core ledger,
+ * since it's necessary to track consistently when tracking Grain distribution.
+ * This type should not grow to include all the data that the UI will
+ * eventually want to show; that should be kept in a different data store which
+ * isn't being used as a transaction ledger.
+ *
+ */
+import {
+  type Uuid,
+  parser as uuidParser,
+  random as randomUuid,
+} from "../util/uuid";
+import * as C from "../util/combo";
+import {
+  type NodeAddressT,
+  NodeAddress,
+  type Node as GraphNode,
+} from "../core/graph";
+
+/**
+ * We validate usernames using GitHub-esque rules.
+ */
+export opaque type Username: string = string;
+const USERNAME_PATTERN = /^@?([A-Za-z0-9-_]+)$/;
```

Comment:
The `@` sign isn’t really part of the username, is it? It’s part of a
username *reference* in some free-form text contexts. But, for instance,
on the main page of this repo GitHub says “decentralion committed … 11
hours ago”, and that contributors include “decentralion” and “wchargin”
and “Beanow”, and that this pull request was opened by “decentralion”—no
`@`-signs anywhere.

I also think that this is confusing because `usernameFromString` accepts
non-canonical inputs and is not injective. As written, we could have a
config file of usernames like

```json
{
  "users": [
    {"username": "@decentralion", "id": "...", "aliases": []},
    {"username": "wchargin", "id": "...", "aliases": []}
  ]
}
```

which would be properly parsed even though the `username` forms are
inconsistent. This makes things like sorting a list of usernames harder.
It also means that logical identity on usernames is not the same as
physical identity, which is unfortunate.

Can we remove the `@` sign here and consistently just use the bare
username?

Reply:
Done